### ISSUE TITLE
More TTS!!

### ISSUE THREAD

**Issue by da3dsoul** [oobabooga/text-generation-webui#885]
Each TTS system has pros and cons.
I'm going to build plugins for each of these to see how they perform, and since they'll be done, I may as well PR them.

https://github.com/coqui-ai/TTS - very good samples
https://github.com/neonbjb/tortoise-tts - also very good samples
https://github.com/CorentinJ/Real-Time-Voice-Cloning - custom voices? looks neat
https://github.com/rhasspy/larynx - very low-spec compatible, acceptable quality
https://github.com/TensorSpeech/TensorFlowTTS - very configurable from what I see

I care less about speed and more about quality, while some people might just want it to run with as little impact as possible.

Umm... we may need to rethink the UI layout for some things if all of these are actually accepted

**Comment by IllogicalDesigns** [comment]
I really like tortise for the quality it produces, but it is slow.  Have you tried this Tortoise-tts-fast repo.  It is no longer in development and I had a hell of a time setting it up, but it is much faster.  https://github.com/152334H/tortoise-tts-fast

**Comment by da3dsoul** [comment]
I'll check it out, thanks

**Comment by Ph0rk0z** [comment]
https://git.ecker.tech/mrq/ai-voice-cloning

**Comment by da3dsoul** [comment]
Ok, Coqui AI TTS is working, at least. More configurability is WIP. It is about as fast as silero using cuda on my A4000. It is slow as balls on cpu (3600X, but I doubt it matters much when it takes 5+ minutes on a recent enough mid-range CPU).

# Compared against Silero
Pros:
- WAY more natural sounding
- Very configurable
- Lots of different models
- Can train custom voices from audio
- Actively in development with some very smart researchers involved (not that Silero isn't)
Cons:
- Speed is probably dependent on hardware. Cuda is required for anything remotely interactive
- Some models seem to have a noisy distortion. Needs more testing
- Can't choose which gpu to use. See https://github.com/coqui-ai/TTS/issues/2282

**Comment by da3dsoul** [comment]
@IllogicalDesigns I've yet to test, but from reading around, it appears that the difference between the main branch and the fast fork is narrowing. Fast is still faster, but only by about 10%, rather than 5-10x like it was. We may need to make a decision between it being fast now and it being maintained

**Comment by St33lMouse** [comment]
So this all runs locally on your machine?  Very interested and will be following keenly. I was thinking of making an extension to use tortoise for voices.

The problem with Silero is quality and you're stuck with the voices they have. The problem with elevenlabs is that everything resides on their machines and they can see what you've been chatting about. Don't like that :) And you have to pay.

**Comment by da3dsoul** [comment]
Yes! That was my intention. I prefer local systems and customization. Coqui is in a PR. Tortoise is next

**Comment by St33lMouse** [comment]
I haven't tried Coqui yet, but I have worked with the MRQ repository. I'm not sure how that differs from the tortoise fast version, but you can certainly do voice training with MRQ. The problem is the speed (of course), and vram usage which will require downgrading your language model if you want everything to fit.

Tortoise or the MRQ version (not sure) has issues with long sentences. MRQ attacked that problem with a new line character to signal breaking off and starting a new generation. That worked, but you tended to get distortion near the ends of lines. Tortoise  may be slow and buggy, but it IS very controllable.

Anyway, good luck on this. I'm dying to try it!

**Comment by St33lMouse** [comment]
Oh, a small but useful feature would be assigning a specific voice to a character automatically.

**Comment by Brawlence** [comment]
You'd probably want to make this into a comparison table, like this

Header1 | Model 1 | Model 2
-|-|-
Type| blabla | haha
Pro| Much | Also many
Con| not many | not much

It's made like this:

```
Header1 | Model 1 | Model 2
-|-|-
Type| blabla | haha
Pro| Much | Also many
Con| not many | not much
```

Also for pros, don't forget to include zero-shot voice cloning. It's very important, for with Tortoise-TTS I can simply put 4 voice files with 40 seconds each to reliably compute voice features and generate (never-before-seen in training) voice.

**Comment by da3dsoul** [comment]
Can confirm.... Tortoise is slow. Got it running in a test. I'll probably try the "fast" version next. It's possible that the first voice I tried happened to be slower than normal. This is the first test, after all.

**Comment by da3dsoul** [comment]
I don't have a 3090. I have an A4000. I mentioned in a previous comment how that compares. I am not using custom voice clips, just the built-in voices. Is ultrafast the preset? I am using standard for preset. I am not using MRQ's fork, though I could potentially. I suppose I can make a branch in my fork and upload what I have. It's still in the testing phase, but so is all of AI, I guess.

**Comment by St33lMouse** [comment]
I've only ever tried the mrq repo. It has a gui with the ultrafast setting. I could look at the preset numbers for you tomorrow--not at my strong computer today. 

Silero for me is very fast. I imagine Coqui will be as well. I'm eager to check out Coqui and see if I can clone voices with it easily enough. If I can, it's golden for me.

**Comment by da3dsoul** [comment]
Based on what I saw, you can. I don't know how good the quality is, but the process seemed simple enough. Here's the branch with the tortoise extension. Run install_tortoise.sh with the current directory set to ./extensions/tortoise_tts, then it should just work.

**Comment by Ph0rk0z** [comment]
Coqui is not good at copying voices like mrq but it is a good sounding TTS. MRQ now uses stock tortoise AFAIK.

Ideally I would like to run TTS/SD on one GPU and the LLM on another.

**Comment by ChristianCCastro** [comment]
Disculpen.  Por no reportarme.    Lo vamos a lanzar. G2.  Pero con nuestra
magia 🪄. Y disculpa no ví tu comentario.  Aquí es para que ganemos todos.
 Pero por favor.   Aboguen por mi.  No me.  Ha caído nada.   Bueno.
Felicitaciones.

Bueno empezamos en hora.      Por favor ocupo.   Alguien. Que represente.
 La.comunidad para que me diga    ya saben  Chelsea       qué r la formula
uno          creo que los gif  del teléfono 📲. Ya se pueden mandar 🚣🫡

**Comment by ChristianCCastro** [comment]
A toda ma..,. Padre.   Ayúdanos. Por qué no tengo idea 💡 Amén




Es broma  somos unos expertos

**Comment by da3dsoul** [comment]
> Ideally I would like to run TTS/SD on one GPU and the LLM on another.

That can be done. Even with the current setup, you can choose which GPU both run on.

@oobabooga can you remove the Spanish nonsense (not because it's Spanish, but because it's nonsense)

**Comment by Ph0rk0z** [comment]
Some AI wanted to make a bug report too xD

**Comment by da3dsoul** [comment]
The tortoise branch has a tortoise_tts_fast extension now. It includes a bash script to successfully install it. I don't have the VRAM to test in UI mode (not using my test scripts), so if one of you would like to, by all means. The fast fork starts its own webui even when called as library for some reason. Since it's not actively developed, we can assume that's just going to be that way.

P.S. this issue got buried fast....

**Comment by da3dsoul** [comment]
I'm dumping some links here, mostly for my own usage. Others may find them useful. They are all related to text normalization, which is the process of expanding text into more proper language or phonemes for TTS. The one from XuanyiZ looks especially neat for something like Neuro-sama reading Twitch chat.
https://github.com/google-research-datasets/TextNormalizationCoveringGrammars
https://github.com/shauryr/google_text_normalization
https://github.com/XuanyiZ/Text-Normalization
https://github.com/csebuetnlp/normalizer
https://github.com/tomaarsen/TTSTextNormalization

**Comment by da3dsoul** [comment]
I merged this into the Coqui PR, because...it's kind of dependent on some core changes (the tts_preprocessor). Now it has Coqui and the Tortoise implementations that have been discussed.

**Comment by da3dsoul** [comment]
@Ph0rk0z @St33lMouse @Brawlence I said it as an aside, but I'm going to tag you. I need testing/assistance on Tortoise. I either need help getting it to run alongside vicuna, which leaves 8GB of VRAM left to work with, and that seems reasonable, or just people to run it and gather data for it.

**Comment by BarfingLemurs** [comment]
@da3dsoul hi, request for recommendations/hints, info on the best backend for a user's setup **within** the interface

**Comment by da3dsoul** [comment]
> @da3dsoul hi, request for recommendations/hints, info on the best backend for a user's setup **within** the interface

I'm not sure what you mean. If you are asking which TTS is the best, then right now, I'd say Coqui, but they each have their own pros and cons. Silero is the easiest to run, as it's equally as fast on CPU, leaving the GPU to run the LLM inferencing. Coqui is higher quality and has a lot of options, but requires some GPU to run. The minimum specs for it depend on settings. The three versions of Tortoise are included because they each have advantages. The original repo presumably will have the most support. The MRQ repo has a lot of extra features. The fast repo is geared towards inferencing as fast as possible. It's no longer maintained, as of a few weeks ago. We'll see if someone picks up the mantle

**Comment by Ph0rk0z** [comment]
All my tortoise and MRQ have different venvs. This forces you to use the same VENV.

If I want to use an AMD or some other GPU for TTS it will not work because the env will be the one from nvidia. Same if mrq or tortoise ever have conflicting requirements for packages.

Also would be nice to have voices automatically read from the voices location instead of being hard coded. If you made other voices in tortoise they won't show up.

**Comment by da3dsoul** [comment]
> All my tortoise and MRQ have different venvs. This forces you to use the same VENV.
> 
> If I want to use an AMD or some other GPU for TTS it will not work because the env will be the one from nvidia. Same if mrq or tortoise ever have conflicting requirements for packages.


Well, the only alternative is calling it via a WebAPI. I don't know if one is built-in. All I saw was a webui. My goal here was to make the usage as simple as possible, like Silero is.


> Also would be nice to have voices automatically read from the voices location instead of being hard coded. If you made other voices in tortoise they won't show up.

A configurable voices path is easy enough. I can add that

**Comment by Ph0rk0z** [comment]
It is indeed working.. but I think tortoise is way heavy for this type of thing on one gpu. Still it's fucking funny to have joe biden responding, even if it does take 60 sec to make it sound good. 

I'll have to see how it juggles loading and unloading a bigger model. I've been doing the 7b so far.

**Comment by da3dsoul** [comment]
Updated to pull the model dir from the args and added settings for the voices dir and sentence length. I'm curious how adjusting the length will affect it. My GPU doesn't even have 18GB of VRAM lol

**Comment by Ph0rk0z** [comment]
I know on MRQ there is some option that makes it use lower vram and other stuff. When I'm doing reggo voice cloning it only uses 6-8gb.

**Comment by da3dsoul** [comment]
Ok, Coqui VC path setting added, along with a bunch of other settings and some fixes

**Comment by Ph0rk0z** [comment]
It does indeed work.. I set the path manually, was the path supposed to go into the speaker box or another one?

**Comment by da3dsoul** [comment]
No, it doesn't have UI yet. A lot of the things I just added need UI. That's the next thing I'll do, probably tomorrow (9PM ATM)

**Comment by St33lMouse** [comment]
OK, time to get this thing installed.  First problem:  running the tortoise install scored me a tortoise folder with a hidden .git folder and nothing else in the tortoise folder. no requirements.txt, nothing.

I'm pretty dumb about this stuff, which makes me a useful tester, I think. Why shouldn't the user just git clone https://github.com/neonbjb/tortoise-tts ?

March me through this like the perfect idiot I am. I'm sure I'll stumble upon ever single unexpected barrier to installation and running the thing. But I want to see this work, so I'll stick with it.

**Comment by Brawlence** [comment]
@da3dsoul 
>I said it as an aside, but I'm going to tag you. I need testing/assistance on Tortoise. I either need help getting it to run alongside vicuna, which leaves 8GB of VRAM left to work with, and that seems reasonable

I implemented [temporary model unloading](https://github.com/Brawlence/SD_api_pics/wiki/VRAM-management-feature) in `sd_api_pictures` specifically to combat this. Additional 4-10 seconds to shuffle models around VRAM and RAM pale in comparison to Tortoise-TTS's innate 60+ seconds generation time on fine presets. 


The up-front cost is pretty high, though. For this to work one needs to implement an API to unload and reload models in Tortoise-TTS (here's [my pull for A1111](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/sd_api_pictures/script.py#L36) for reference). Then, in the extension scripts:

```python
from modules.models import reload_model, unload_model
```
implement something akin to [give_VRAM_priority()](https://github.com/oobabooga/text-generation-webui/blob/b57ffc2ec9bc32ff0cc82ba97edf4986ecf426fb/extensions/sd_api_pictures/script.py#L36) function
```python
def give_VRAM_priority(actor):
    global shared, params

    if actor == 'TTS':
        unload_model()
        print("Requesting Tortoise to re-load voice generation model...")
        response = requests.post(url=f'{Whats the api to reload model in Tortoise?}', json='')
        response.raise_for_status()

    elif actor == 'LLM':
        print("Requesting Tortoise to vacate VRAM...")
        response = requests.post(url=f'{Whats the api to unload model in Tortoise?}', json='')
        response.raise_for_status()
        reload_model()
 …
```

And then just `give_VRAM_priority(TTS)` every time you want to generate audio.

**Comment by da3dsoul** [comment]
> OK, time to get this thing installed. First problem: running the tortoise install scored me a tortoise folder with a hidden .git folder and nothing else in the tortoise folder. no requirements.txt, nothing.
> 
> I'm pretty dumb about this stuff, which makes me a useful tester, I think. Why shouldn't the user just git clone https://github.com/neonbjb/tortoise-tts ?
> 
> March me through this like the perfect idiot I am. I'm sure I'll stumble upon ever single unexpected barrier to installation and running the thing. But I want to see this work, so I'll stick with it.

It uses a sparse checkout to skip the unnecessary parts. I had issues getting it to work on Windows. I'm not sure why it wouldn't work. Needs looking into. I'm making settings UI for stuff, and Bawlence posted that nice bit of kit for me to look at, too, so I'll definitely be working on stuff today.

If you'd like to just get it running, the first line of the install_tortoise.sh can be run without the double hyphenated options and it should do a full clone

**Comment by St33lMouse** [comment]
Note: all this is with the conda environment active.

OK, yeah, that script gives me trouble. Did a straight up git clone:

git clone -n https://github.com/neonbjb/tortoise-tts tortoise but this fails to do the clone because of the rename, I think.

So now: git clone https://github.com/neonbjb/tortoise-tts, rename tortoise_tts to tortoise, cd tortoise, python -m pip install -r ./requirements.txt

This runs through the install, but fails with: 
note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

Tried to run anyway, loaded chat mode and the tortoise_tts extension, reloaded the gui:

To create a public link, set `share=True` in `launch()`.
Closing server running on port: 7860
Loading the extension "gallery"... Ok.
Loading the extension "tortoise_tts"... Fail.
Traceback (most recent call last):
  File "/media/mouse/eastwatch/text-generation-webui/modules/extensions.py", line 19, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "/media/mouse/eastwatch/text-generation-webui/extensions/tortoise_tts/script.py", line 8, in <module>
    from .tortoise.tortoise import api
  File "/media/mouse/eastwatch/text-generation-webui/extensions/tortoise_tts/tortoise/tortoise/api.py", line 9, in <module>
    import progressbar
ModuleNotFoundError: No module named 'progressbar'
Running on local URL:  http://127.0.0.1:7860

Frustrating!

Went back to the tortoise directory:  pip install progressbar, run server.py again.  This time it wants rotary_embedding_torch.  OK.  pip install and run server.py again. gui loads in default mode. Switch to chat and select the extension, but gui won't load. No error messages in the terminal.

Finally I reloaded server.py, opened the gui and switched to chat mode without the extension. no problem. switched to chat mode with the extension, gui won't load, no errors in the terminal.

My guess is that tortoise had a bad install way back at the top when it complained about meta generation failed. I fear renaming the folder tortoise was not a good thing, but I'm not sure. This ends my first attempt. I'm going to try and install tortoise in the extensions folder using the tortoise installation instructions and see if I can avoid errors that way.

**Comment by St33lMouse** [comment]
Round two:  I installed tortoise and got it to work by itself. Had to downgrade numpy and install a few more packages that didn't install themselves for whatever reason. I still got the meta data error, but I was able to run the tortoise script and got it to generate some voiced text for me. I'm getting better at installing this stubborn ai stuff...

Having got tortoise to work by itself, I renamed the tortoise tss to just tortoise so your script could find it, then ran server.py, switched to chat mode and loaded your extension. No luck--gui won't connect.

I went back to tortoise to make sure it would generate a clip with its directory renamed to just tortoise to make sure it worked, and it worked fine. So no problem there, although I'm still uncertain about that rename. Could be asking for trouble.

The only thing I've got left is to figure out what that metadata error was all about, but tortoise works in spite of that error, so I'm pretty much at a dead end with tortoise tts. May try the MRQ version shortly. 

I notice that coqui extension has no install script, and requirements.txt simply says tts in it.  Should I try to git clone coqui in that folder?

**Comment by da3dsoul** [comment]
Coqui only needs pip install TTS, as it's on pip.

Try the tortoise clone with
```
git clone https://github.com/neonbjb/tortoise-tts.git tortoise
cd tortoise
pip install -r ./requirements.txt
```
```

**Comment by St33lMouse** [comment]
OK, that worked this time. No idea why it failed last time. I note that tortoise has another step:  python setup.py install which I went ahead and ran from inside tortoise. It then wanted a few more dependencies, easily installed.  I tested a voice generation with tortoise by itself, and it worked.

Trying to load your extension in the gui fails, and trying to load it with python server.py --chat --extensions tortoise_tts

gets this error (and then the gui loads anyway)

(testgen) mouse@mousehole:/media/mouse/eastwatch/experiment/text-generation-webui$ python server.py --chat --extensions tortoise_tts
bin /home/mouse/anaconda3/envs/testgen/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cuda117.so
Loading the extension "tortoise_tts"... Fail.
Traceback (most recent call last):
  File "/media/mouse/eastwatch/experiment/text-generation-webui/modules/extensions.py", line 19, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "/media/mouse/eastwatch/experiment/text-generation-webui/extensions/tortoise_tts/script.py", line 94, in <module>
    model, voice_samples, conditioning_latents = load_model()
  File "/media/mouse/eastwatch/experiment/text-generation-webui/extensions/tortoise_tts/script.py", line 87, in load_model
    tts = api.TextToSpeech(models_dir=os.path.join(models_dir, 'tortoise'), device=params['device'])
  File "/media/mouse/eastwatch/experiment/text-generation-webui/extensions/tortoise_tts/tortoise/tortoise/api.py", line 231, in __init__
    self.autoregressive.load_state_dict(torch.load(get_model_path('autoregressive.pth', models_dir)))
  File "/home/mouse/.local/lib/python3.10/site-packages/torch/serialization.py", line 771, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/home/mouse/.local/lib/python3.10/site-packages/torch/serialization.py", line 270, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/home/mouse/.local/lib/python3.10/site-packages/torch/serialization.py", line 251, in __init__
    super(_open_file, self).__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: 'models/tortoise/autoregressive.pth'

**Comment by Ph0rk0z** [comment]
Its supposed to download that stuff the first time you run it. Did you load the extension from the command line?

**Comment by St33lMouse** [comment]
You mean this here: python server.py --chat --extensions tortoise_tts  will pause and download that stuff?  My internet is kind of sticky today, so it must have timed out.  I'll try again.

**Comment by Ph0rk0z** [comment]
Yea.. and from the UI it won't.

**Comment by da3dsoul** [comment]
I'll look at it.
@St33lMouse the difference in the commands is `-n` which means no checkout. You don't want to run `setup.py install` mostly because that will install it to the conda environment, and it might cause package conflicts if you wanted to try each of the versions of tortoise side by side. The extension loads tortoise from the folder it cloned, rather than via an installed package

**Comment by St33lMouse** [comment]
If I have to do another reinstall I'll avoid the setup.py mistake.  OK, got coqui interface to load. I loaded the example character and was surprised to hear a voice start talking to me right away! 

I don't see any folders or support files or anything. How do you change voices and stuff? Speaker and language boxes in the UI are empty.

It's getting late for me, so I'll be back in another 10 hours or so and try and get tortoise going. We should pay attention to my little installation problems here and spell things out clearly for other users. After months of installing AI things like this I have some idea about what to do, and I still run into lots of problems. Others will be completely clueless. Even things you think are obvious are not at all obvious to newbies, and they'll either bug you with questions or just give up and go away.

**Comment by da3dsoul** [comment]
All of it needs more UI work. You can use the settings in the Python file and presumably the built in settings that I haven't figured out yet.
I've been trying to figure out the gradio UI framework callbacks

**Comment by St33lMouse** [comment]
I'm still getting this error when I try to load the tortoise extension from server.py from the terminal:

FileNotFoundError: [Errno 2] No such file or directory: 'models/tortoise/autoregressive.pth'

The error comes up very quickly. I don't think it's trying to download any files. How can I get that missing stuff downloaded?

**Comment by Ph0rk0z** [comment]
Try to run the tortoise manually with stuff from it's repo. That will fix the error. See I used tortoise beforehand and those files all downloaded.

**Comment by St33lMouse** [comment]
OK...today's installation efforts...success! I finally realized that when it was asking for those files at models/tortoise/autoregressive.pth'

it actually meant it wanted a tortoise folder in the text-generation-webui/models folder, NOT in the tortoise installation. I don't see how those files could get downloaded automatically, so I dredged them out of my MRQ installation and put them there. That actually worked! But this folder structure is going to give other people problems, just like me. At the very least, they'll have to be told where to put those files.

So now the extension loaded, and it...worked!  Slow as a tortoise, of course. The initial assistant greeting is long and took 400 seconds to generate. I set it to ultra fast and it took 2 minutes to generate a 14 second clip. Are you sure the speed settings are working?  

I like how you print the text and also show the sound file. Personally, I want to see that text.  

Memory usage is heavy. I have pyg6B loaded, and tortoise pushes me close to the edge. Nothing you can do about that, I figure. My plan is to use alpaca 13B 4bit, which uses about 8gb vram, I think, that should be a good balance and room for safety.

Voices could use a refresh button in case the user drops more voices in the folder during a session. Probably not a high priority thing.

Next up, I'm going to try and get tortoise fast and mrq working. I'm not sure you need them both. As a matter of fact, you might just want to go with tortoise fast and drop the other two. MRQ's purpose is voice training, though, And a trained MRQ voice is pretty cool.  I'll explore this more carefully.

**Comment by St33lMouse** [comment]
By the way, I'm running KDE ubuntu on a 3090 if you need that info.

Now trying to get mrq working. python server.py --chat --extensions tortoise_tts_mrq:
ModuleNotFoundError: No module named 'tortoise.models.bigvgan'

Can't figure this one out. Need to know exactly what files it is looking for and where they should go.

And finally tortoise fast gets me:
ImportError: cannot import name 'VocConf' from 'tortoise.models.vocoder' (/home/mouse/anaconda3/envs/soundgen/lib/python3.10/site-packages/TorToiSe-2.4.2-py3.10.egg/tortoise/models/vocoder.py)

**Comment by da3dsoul** [comment]
> Dropping a new voice into the voices folder does not update the list of possible voices, I think.

That is WIP and almost done. I figured out the issue I had with UI updates, so I can do a bunch now. 

> Memory usage is heavy. I have pyg6B loaded, and tortoise pushes me close to the edge. Nothing you can do about that, I figure.

I'm going to add the code that Brawlence mentioned before that will allow swapping the models out. I've also added the low VRAM setting to MRQ.

These changes are all still local. The next update will be a fairly big one

**Comment by St33lMouse** [comment]
OK. remember, I can't get mrq or fast to work right now, only regular tortoise per my last report.

**Comment by St33lMouse** [comment]
Yes, same procedure for tortoise. Everything installed fine, no errors.

**Comment by St33lMouse** [comment]
OK, nuked the conda environment and reinstalled, did not run setup.py. Installed Tortoise and tortoise mrq. Tortoise extension runs fine, but mrq complained about emma (the default tortoise voice), so I copied that voice folder to a couple of likely locations in the MRQ tree and the extension finally ran correctly.  

The speed is still a couple of minutes at ultra fast (depending on the size of the sentence), and it shouldn't be--A quick sentence at ultra fast is about 15 seconds on my regular mrq installation. I note that the terminal shows it is computing conditional latents. I think mrq does that only once and saves the results in the voice folder so it doesn't have to compute every time.

Anyway, got it running, looking forward to your updates.

**Comment by da3dsoul** [comment]
Yep I've made decent progress. I'm testing at the moment. I forgot MRQ doesn't come with the voices. I can maybe make a script to check if they exist and pull them. That's a comparatively low priority, though.

This next update will have a lot of settings and UI for them, including custom directories for relevant things.

**Comment by St33lMouse** [comment]
No big deal on the pre-made tortoise voices for now, I agree.

The way MRQ works is you can just drop voices into the MRQ voice folder and it will do a very good imitation of them. it is a key feature--THE key feature and reason to use it. I doubt that MRQ is any faster that tortoise_fast, so this is the feature I'm hoping to see here.

Secondarily, MRQ allows voice training, and it produces a voice model. In MRQ you can point the gui at your trained model instead of autoregression.pth and use your voice clips to produce a high quality cloned voice. Training is beyond the scope of what you're doing, but being able to choose a model trained in the full MRQ setup would preserve what MRQ is all about.

**Comment by da3dsoul** [comment]
I see. I might change things up a bit to allow changing the directory from the UI. I made it obey the model dir argument, but I may want to choose a subdirectory structure to load into a dropdown

**Comment by St33lMouse** [comment]
Have you installed the full MRQ to see how it operates?  He moved the folders around from the original tortoise setup, so watch out for that. But all you gotta do to make a new voice is take a few 10 second clips in wav format, put them in a folder in the voices folder, and start generating. Remember that part of the speed comes from an initial conditional latents  calculation that he saves in the same folder as the wav files.

**Comment by da3dsoul** [comment]
I have not tried to use it past generating speech in the same way as the base tortoise. My goal was getting it running, letting people give feedback, and iterate. So far, according to plan lol.
I noticed the code was reorganized a bit. I worked around that, too.

**Comment by da3dsoul** [comment]
Update is pushed. It's not perfect, and not all of the things mentioned are implemented yet. I'm trying to figure out the speed. It's very slow, and I'm not sure if it's my fault or with my MRQ setup. I haven't generated latents, for example

**Comment by St33lMouse** [comment]
It's probably calculating them every single generation instead of just once. So I just do a git pull, or something else?

**Comment by da3dsoul** [comment]
Yeah a `git fetch --all` and `git pull` should do it. If you get merge issues, you can use `git stash`

**Comment by St33lMouse** [comment]
OK, updated no problems. Speed seems pretty good at ultrafast. Probably should make sure ultrafast is your default choice, since what will happen is players will pick the example character which has a very long intro at standard generation, and then they'll be in for a ten minute sound generation treat :) They'll figure it's broken.

What's the purpose of the custom voice folder, and how might I use it?

By the way, I'm not a coder, I'm a game designer. My technical skills are no match for your own. I use, I don't make :)

**Comment by da3dsoul** [comment]
Custom directories are basically because I have a small boot SSD and a large bcached raid, so I point those to my mass storage. It also allows you to share a voices and models folder between the different builds

**Comment by da3dsoul** [comment]
When you say the speed is good...do you mean like 5 minutes or 20 seconds?

**Comment by St33lMouse** [comment]
45 seconds for a short generation at ultrafast. Double for standard. 

The models are getting loaded and unloaded to save vram?  I turned off the vram saving options, saw vram usage spike to an incredible 18gigs, and I only have the little opt-1.3 model loaded to prevent oom.

**Comment by da3dsoul** [comment]
Yeah MRQ seems to use a boatload of VRAM. I added all kinds of knobs to mess with, matching a lot of the options it has in the API. I needed to use the model unloading and low vram mode just to run the thing.

After building the latents, it dropped to about 1:45s for a test string of 11s. That actually matches more or less what you are getting with the difference of A4000 vs 3090. I can look at more stuff tomorrow. It is late

**Comment by St33lMouse** [comment]
Using the custom voice directory box works well. no need to refresh, it just looked in the new directory.  Access to all my voices now. 

OK, talk to you later.

**Comment by da3dsoul** [comment]
Ok, most of those options (the ones that are supported) from MRQ are now added to the other two. The fast branch is once again good, and it's fast enough to be interactive. With the model swapping, you can run it on a single reasonable GPU, as well!
Tortoise does have its own issues with quality, but we may be able to work around some with preprocessing.
In this example, there are some weird cadences and pauses:
> Waffles and cakes are two different things, although they may share some similarities. Waffles are made with batter or dough that's poured into a grid-like iron and cooked until crispy on the outside and fluffy on the inside. Cakes, on the other hand, are baked in a rectangular pan and typically have a more moist and dense texture. While both waffles and cakes can be sweet, they are not the same thing.

https://user-images.githubusercontent.com/5205810/233737509-af7db745-7902-4363-bcde-c00c96a6af8b.mp4

**Comment by Ph0rk0z** [comment]
Did you get a look at bark? https://github.com/oobabooga/text-generation-webui/issues/1423

**Comment by da3dsoul** [comment]
I did not. I'll take a look. My next step is improving the oobx (out of box experience) for this, and then that seems like a decent thing to look at next.

**Comment by wsippel** [comment]
@Ph0rk0z you could just use the extension that already exists? It's in the bug report you linked to, you know...

**Comment by da3dsoul** [comment]
@wsippel sweet! Feel free to use anything in here if it helps make it better

**Comment by bubbabug** [comment]
Hello de3dsoul, I am very interested in getting coqui_tts working, but I am having trouble. I have downloaded and copied over the extensions folder and pip installed the TTS requirement, but when I try to run the extension I get a failure to find tts_preprocessor file in modules. I attempted to copy over the preprocessor file from silero and I was able to startup, but it throws an error on generation. Where do I get the approrpiate tts_preprocessor file?

**Comment by da3dsoul** [comment]
@bubbabug yeah, I had some basic instructions at https://github.com/oobabooga/text-generation-webui/issues/885#issuecomment-1509513725

**Comment by wsippel** [comment]
@da3dsoul Thank you, I'll take a look! Bark is quite unusual in that it does almost everything on it's own and exposes next to nothing. It's transformer-based, so it's fundamentally different from most other TTS systems. It generates speech the same way LLMs generate text, by constantly guessing the next phoneme - it only uses the input text as guidance (which means it can go off the rails and start to ramble, ignoring most or all of the input text). It also generates ambient sounds. I asked it to tell my about Vicunas, and it did, but with jungle noises in the background. It also takes directions, like a movie script. Super interesting and powerful, but also very different from any other TTS engine I've ever worked with. Extremely fun to explore.

**Comment by da3dsoul** [comment]
Oh that's interesting. It's maybe not the best for a professional TTS system (where accuracy is king), but definitely neat for creative applications

**Comment by wsippel** [comment]
Yeah, it's very unpredictable, though it can be tamed to a degree. Maybe not really ideal for a digital assistant, let alone a screen reader, but the potential for stuff like an AI storyteller or roleplay is incredible. It 'acts'. Even at lower temperatures (the only controls it provides, just like other generative AI models), it might do stuff like clearing its throat in the middle of a sentence, stutter, or inject dramatic pauses, 'um's, 'ah's or 'ya know's that aren't in the original text. You should give it a spin since you're interested in TTS - it's a trip.

**Comment by St33lMouse** [comment]
I'm not too sure that selecting  a different voice is working in the MRQ version. I tried a male voice, but it still came out female. Might just be using emma for everything.

Opened my main MRQ install to have a look at default settings: cwp weight 1, diffusion temp: 1, length penalty 1, repetition penalty 2, conditoning free k: 2, temperature 0.8, diffusion model: diffusion_decoder.pth, vocoder:bigvgan_24khz_100 band

Using a custom finetune model (instead of autoregressive.pth) on the standard MRQ install with the voice files used to make the finetune produced noticeably recognizable results in 4 seconds at ultrafast, very acceptable. This was the text:

This is a test of the emergency broadcasting system. It is only another test.

vram started at 6 gigs, spiked only 1 more gig during that generation.

So the massive vram spikes you're experiencing isn't what I'm seeing in the regular install. At standard I watched it spike by 5 gigs, which isn't a good thing, but I get the same kind of spikes when doing text generations.

**Comment by da3dsoul** [comment]
@St33lMouse there's a bug that I addressed in https://github.com/oobabooga/text-generation-webui/pull/1484. Basically, settings aren't loaded before init, so the voices aren't properly loaded at init. You can refresh it by just adding a / to the voice dir and deleting it

**Comment by St33lMouse** [comment]
I don't understand. what voice dir? Put a slash where?

Right now I can load plain tortoise_tts with python server.py --chat --extensions tortoise_tts
And I get a working tortoise interface with a list of all the tortoise voices. Selecting a new voice changes to the new voice correctly. I test first with a woman's voice, then a man's to be sure it has changed to a new voice. The custom voices box in the interface is empty.

Adding a folder path to the custom voices box will immediately add those voices to the list, and selecting them works. I don't think it is updating quality settings when I move from ultra fast to standard, because it still generates very quickly.

So basically, the plain tortoise_tts extension is working pretty good.

Today, I'm not having much success with the MRQ version. I'm not able to get it to work well enough to test switching voices. I am ooming with just a tiny opt 1.3B model loaded and 24 gigs of vram. When this happens, it passes the message back to my character, as though the character said it:

CUDA out of memory. Tried to allocate 16.02 GiB (GPU 0; 23.70 GiB total capacity; 18.06 GiB already allocated; 4.04 GiB free; 18.09 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation. See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

**Comment by da3dsoul** [comment]
Ok, sounds like you weren't affected by the bug I mentioned. I can test further with the preset and tuning settings, but they were working.
How do you run it in a normal install? I pulled my method from read.py

**Comment by St33lMouse** [comment]
How do you run what?

Hang on, I just realized I didn't do a git pull, so I just did a git pull for your repo, so gotta test again.

**Comment by da3dsoul** [comment]
I mean when you run it without text-generation-webui, what command do you use?

**Comment by St33lMouse** [comment]
python server.py --chat --extensions tortoise_tts_mrq

After the git pull, tested again. MRQ is still ooming for me, so can't check switching voices. regular tortoise is fine, but I've seen it spike by 14 gigs vram. Still don't think quality settings in regular tortoise are doing anything.

**Comment by tanfarou** [comment]
> @Ph0rk0z @St33lMouse @Brawlence I said it as an aside, but I'm going to tag you. I need testing/assistance on Tortoise. I either need help getting it to run alongside vicuna, which leaves 8GB of VRAM left to work with, and that seems reasonable, or just people to run it and gather data for it. It is [here](https://github.com/da3dsoul/text-generation-webui). If you just want to take my changes, you can grab `modules/tts_preprocessor.py` and the `extensions/tortoise_tts*` folders. Each one has an `install_tortoise.sh` script. cd into the `extensions/tortoise_tts*` folder and run `sh install_tortoise.sh`. The scripts will override the version bindings to not fuck your conda environment. That much is tested.

----
I am testing on a Windows PC running 3090. Would the install_tortoise.sh create a tortoise folder and then clone everything into directly, without adding "tortoise_tts_fast" again? (sorry, not familiar with shell). If I can get it running, can do some tests with vicuna running.

**Comment by da3dsoul** [comment]
It should run with git bash, but WSL is recommended for text-generation-webui on windows, as far as I've seen

**Comment by da3dsoul** [comment]
> So the massive vram spikes you're experiencing isn't what I'm seeing in the regular install. At standard I watched it spike by 5 gigs, which isn't a good thing, but I get the same kind of spikes when doing text generations.

I thought you meant that running MRQ standalone didn't have the VRAM issues, but after rereading, you could have meant the non-mrq version

**Comment by St33lMouse** [comment]
I have never seen stand alone tortoise or mrq spike through the roof. They both seem to use about 6 gigs for inference. That doesn't mean they never to, but I haven't seen it.  Yours versions--both tortoise and mrq-- I've seen spike by more than 14 gigs.

**Comment by Ph0rk0z** [comment]
I think they both have parameters that need to be passed for that. Otherwise they use all the vram they can.

**Comment by da3dsoul** [comment]
There is a "Low VRAM" mode for both, but that doesn't do that. It just doesn't precache. It only uses as much VRAM as it "needs", but that's still a lot. I've not tested standalone vs this. The preset definitely makes a difference in VRAM, and MRQ seems to use the most VRAM of all of them.
I fixed the issue with the parameters in the UI throwing an error, too

**Comment by da3dsoul** [comment]
@tanfarou I'm not sure I understood what you meant. The install.sh just runs a few git commands, edits requirements.txt to not cause version conflicts, then runs pip install -r requirements.txt.
The fancy git commands are for a sparse checkout, which only pulls what we need, rather than the whole repository. The goal was to avoid wasting space on things like the example output audio files

**Comment by oobabooga** [comment]
I'll write some comments for #1106 here:

* coqui-tts sounds better than Silero, which is too unemotional. I liked it, but a con was that it installed a huge number of dependencies. It would also be nice to have more default voices to select from a dropdown menu, is that possible?
* I couldn't test any of the tortoise implementations because all the install script did was create an empty folder with a `.git` in it as someone reported above. From past experience, it uses a huge amount of VRAM and is insanely slow. Given this and given the complicated installation procedure, I vote for not including it as a built-in extension into the repository.
* For bark, there are these two extensions and I haven't tried any of them:

https://github.com/minemo/text-generation-webui-barktts
https://github.com/wsippel/bark_tts

**Comment by da3dsoul** [comment]
> I couldn't test any of the tortoise implementations because all the install script did was create an empty folder with a .git in it as someone reported above. From past experience, it uses a huge amount of VRAM and is insanely slow. Given this and given the complicated installation procedure, I vote for not including it as a built-in extension into the repository.

Yeah it has high minimum requirements, for sure. The fast repo is actually fast, especially if you pre-compute the conditioning latents. I got it to unload the LLM, load tortoise, run it, and reload the LLM in under 30s for some responses.
I didn't want to just clone the whole repo, but it might be necessary considering how inconsistent the sparse checkout is.

If it doesn't make it in as an included extension, we should probably have a wiki list with other extensions people have made. I didn't know about a lot of them.

**Comment by da3dsoul** [comment]
> It would also be nice to have more default voices to select from a dropdown menu, is that possible?

I'll look into that today if I find time. Tortoise got most of the UI love, as it was harder to work with and got more feedback.

**Comment by da3dsoul** [comment]
Ok, I added some descriptions to the model dropdown and made note of which ones need espeak-ng for Coqui. All of the English models are added. Due to how many models there are, a different kind of UI is necessary for all of the different languages.

**Comment by ksylvan** [comment]
I'm very interested in this. What's the status of this work?

**Comment by FerLuisxd** [comment]
Could also add the edge tts that is free: https://pypi.org/project/edge-tts/

**Comment by ChristianCCastro** [comment]
santa cachucha de las cachuchas

**Comment by Cloudwalker2k3** [comment]
> Ok, I added some descriptions to the model dropdown and made note of which ones need espeak-ng for Coqui. All of the English models are added. Due to how many models there are, a different kind of UI is necessary for all of the different languages.

Now matter how I try to get TTS by MRQ working with the web ui, there is always some failure. It can't find the TTS Processor. Unable to find Modules or even Tortoise.

da3dsoul, is there a process you use to get this functionality working with the oobabooga web ui?

**Comment by Cloudwalker2k3** [comment]
> > da3dsoul, is there a process you use to get this functionality working with the oobabooga web ui?
> 
> I have an open PR, which I assume you are trying to use. As part of that PR, tts_preprocessor was moved. Check the changes tab. There's also an install script. You may need to install the original Tortoise and grab the voice models from there. Tortoise_Fast has that issue iirc

I had this error come up:

**Comment by da3dsoul** [comment]
Presumably the branch became out of date. I'll update it

**Comment by Cloudwalker2k3** [comment]
> Presumably the branch became out of date. I'll update it

Just checking in, are you still looking into getting Tortoise TTS fast or MRQ updated?

**Comment by da3dsoul** [comment]
I forgot. Hopefully I'll remember in the morning, as it is quite late

**Comment by da3dsoul** [comment]
Updated

**Comment by Ph0rk0z** [comment]
For extra fun.. apply an RVC model over TTS output. It might be faster and better to use some janky but QUICK tts like coqui and then throw a proper RVC (or even SVC) over it.

That way instead of loooong generations from tortoise you just have a very fast TTS + short pitch fixing inference but still get a "clone" that talks like the character.

Just an idea.

**Comment by CRCODE22** [comment]
coqui_tts extension does not work anymore with latest text generation web ui anyone know how to fix it?

2023-09-28 22:24:41 ERROR:Failed to load the extension "coqui_tts".
Traceback (most recent call last):
  File "K:\text-generation-webui\modules\extensions.py", line 36, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 7, in <module>
    from modules import chat, shared, tts_preprocessor
ImportError: cannot import name 'tts_preprocessor' from 'modules' (unknown location)

**Comment by da3dsoul** [comment]
I would not be surprised if it takes more than this, but that file is moved to the modules folder in the PR.

**Comment by CRCODE22** [comment]
I fixed some of the other issues but now these remain that I have been unable to fix yet.

2023-09-29 01:29:13 ERROR:Failed to load the extension "coqui_tts".
Traceback (most recent call last):
  File "K:\text-generation-webui\modules\extensions.py", line 36, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 71, in <module>
    model, speaker, language = load_model()
  File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 56, in load_model
    tts = TTS(params['model_id'], gpu=params['Cuda'])
TypeError: 'module' object is not callable

**Comment by da3dsoul** [comment]
Can you gist your updated script.py and link it? I'll take a look and see if I can spot the error

**Comment by chigkim** [comment]
Has anyone tried incorporating Piper? Extremely fast and decent quality!
https://github.com/rhasspy/piper

**Comment by Xanw1ch** [comment]
> I fixed some of the other issues but now these remain that I have been unable to fix yet.
> 
> 2023-09-29 01:29:13 ERROR:Failed to load the extension "coqui_tts". Traceback (most recent call last): File "K:\text-generation-webui\modules\extensions.py", line 36, in load_extensions exec(f"import extensions.{name}.script") File "", line 1, in File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 71, in model, speaker, language = load_model() File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 56, in load_model tts = TTS(params['model_id'], gpu=params['Cuda']) TypeError: 'module' object is not callable

I had the same error. It's specifically for Coqui-AI. It appears there's another dependency 
text-generation-webui-main\\installer_files\\env\\Lib\\site-packages\\TTS\\__init__.py

that TTS refers to -- AFTER def load_model(): is called in script.py for the extension.

My solution was to use "as" for the import: from TTS.api import TTS as Coqui_TTS #line 10

And then call tts = Coqui_TTS(params['model_id'], gpu=params['Cuda'])  #In line 56

Seems to work that way for me.

**Comment by CRCODE22** [comment]
> Can you gist your updated script.py and link it? I'll take a look and see if I can spot the error. Edit: you can also PR against my fork if you want to contribute to updating it

Sorry for my slow reply I just saw your reply I have forked your repository and I will try to get it updated to use the latest Oobabooga original repository since yours is outdated and see if I can make it work. It would however help if you could sync yours with the original repository that way people like me can do our best to solve problems. I would love for Coqui TTS to work with the latest OobaBooga.

**Comment by CRCODE22** [comment]
> > I fixed some of the other issues but now these remain that I have been unable to fix yet.
> > 2023-09-29 01:29:13 ERROR:Failed to load the extension "coqui_tts". Traceback (most recent call last): File "K:\text-generation-webui\modules\extensions.py", line 36, in load_extensions exec(f"import extensions.{name}.script") File "", line 1, in File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 71, in model, speaker, language = load_model() File "K:\text-generation-webui\extensions\coqui_tts\script.py", line 56, in load_model tts = TTS(params['model_id'], gpu=params['Cuda']) TypeError: 'module' object is not callable
> 
> I had the same error. It's specifically for Coqui-AI. It appears there's another dependency text-generation-webui-main\installer_files\env\Lib\site-packages\TTS\**init**.py
> 
> that TTS refers to -- AFTER def load_model(): is called in script.py for the extension.
> 
> My solution was to use "as" for the import: from TTS.api import TTS as Coqui_TTS #line 10
> 
> And then call tts = Coqui_TTS(params['model_id'], gpu=params['Cuda']) #In line 56
> 
> Seems to work that way for me.

I have made a fork I do not understand exactly what you wrote could you make the changes within my fork? That way I can see what you mean and if it works.

Thank you for replying :)
